### PR TITLE
exposing default comparator

### LINKED ISSUE #28: Expose direct access to comparator function.
Simple sync API to run the comparator against given inputs. Will be most useful when the comparator is properly swappable.

### PR DESCRIPTION
**Pull request by deepak1556**
Closes #28 

Would like to generalise this however for swappable comparators, i am thinking some api like

```
db.compare('comparatorname', ....) //module providing a list of custom comparators for users to choose
```

Also where do the test for this go, here or in `abstract-leveldown` ?

### CODE REVIEW

File: leveldown.js [hunk 1]
```diff
@@ -57,6 +57,9 @@ LevelDOWN.prototype._approximateSize = function (start, end, callback) {
   this.binding.approximateSize(start, end, callback)
 }
 
+LevelDOWN.prototype.compare = function (key, value) {
```
Comment (juliangruber) [inline leveldown.js]: should be `keyA`, `keyB` or something like that, since we're comparing two keys

Comment (deepak1556) [inline leveldown.js]: oops, yeah. Thanks!

**Comment by juliangruber** [comment]
lgtm, but we should add this to abstract-leveldown as well. other thoughts?

**Comment by deepak1556** [comment]
yup that should be done, but one doubt though the default comparator returns `0,1,-1` . Should this be made standard for users implementing `_compare` in `abstract-leveldown` ?

**Comment by juliangruber** [comment]
i'd make it the same as `Array#sort`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

**Comment by rvagg** [comment]
looks fine to me, nice and clean implementation and something that's been talked about for a while now. My only concern is the sync vs async question as per rvagg/abstract-leveldown#51

**Comment by ralphtheninja** [comment]
lgtm

**Comment by ralphtheninja** [comment]
@deepak1556 I'd like to revive this. Mind rebasing it onto latest master?

**Comment by deepak1556** [comment]
@ralphtheninja have updated the branch and have also updated the corresponding PR in https://github.com/Level/abstract-leveldown/pull/51

**Comment by ralphtheninja** [comment]
> @ralphtheninja have updated the branch and have also updated the corresponding PR in Level/abstract-leveldown#51

Nice! There seems to be some problems compiling the code though.

**Comment by deepak1556** [comment]
@ralphtheninja sorry about that, forgot to test in other versions. Fixed, thanks!

**Comment by ralphtheninja** [comment]
> @ralphtheninja sorry about that, forgot to test in other versions. Fixed, thanks!

Forgot to push your changes? :)

**Comment by deepak1556** [comment]
PR was updated, the compilation problem seems fixed. Only the failing tests due to missing abstract leveldown [PR](https://github.com/Level/abstract-leveldown/pull/51) ?

**Comment by ralphtheninja** [comment]
> PR was updated, the compilation problem seems fixed. Only the failing tests due to missing abstract leveldown PR ?

Aaah ok. I'll check it out .. tomorrow :)

**Comment by ralphtheninja** [comment]
Once https://github.com/Level/abstract-leveldown/pull/51 has been merged and released we can release a new *minor* for `leveldown`. Long time since we did that :)

File: leveldown.js [hunk 2]
```diff
@@ -57,6 +57,9 @@ LevelDOWN.prototype._approximateSize = function (start, end, callback) {
   this.binding.approximateSize(start, end, callback)
 }
 
+LevelDOWN.prototype._compare = function (keyA, keyB) {
```
Comment (deepak1556) [inline leveldown.js]: @ralphtheninja `LevelDOWN.prototype._compare` is only defined here.

File: test/compare-test.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+test('setUp common', testCommon.setUp)
+
+test('setUp db', function (t) {
+  db = leveldown(testCommon.location())
+  db.open(t.end.bind(t))
+})
+
+test('test argument-less compare() throws', function (t) {
+  t.throws(
+    db.compare.bind(db)
```
Comment (juliangruber) [inline test/compare-test.js]: db.compare is now undefined, see travis

**Comment by ralphtheninja** [comment]
Too many things have happened in the node world with new versions and new nan etc. Please open a new PR based on latest master.

**Pull request closed without merging**